Keep input header when converting outer join to inner join

antaljanosbenjamin opened the pull request:
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)


### Changelog entry (a [user-readable short description]

Fix a `LOGICAL_ERROR`s that caused by not wanted modification of query plan when converting outer join to inner join. Also relax the requirements of optimization to be able to apply it in cases when injective functions are applied to the aggregating keys during joins.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->

antaljanosbenjamin committed:
Add basic implementation

antaljanosbenjamin committed:
Add test to verify no logical errors

antaljanosbenjamin committed:
Consider injective functions

With injective functions, we can still guarantee that there is at most one row from the interesting side because the aggregation produces at most one row per aggregation key and injective functions will keep this uniqueness.

novikd commented on src/Processors/QueryPlan/Optimizations/convertOuterJoinToInnerJoin.cpp:
```diff
@@ -99,14 +99,55 @@ size_t tryConvertAnyOuterJoinToInnerJoin(
+    {
+        auto [predicate_op, lhs, rhs] = expr.asBinaryPredicate();
+        if (predicate_op != JoinConditionOperator::Equals)
+            continue;
```
Yes, you are right. What we want to check here is the size of the matching group of the interesting side. We can estimate it only for equality predicates.

novikd commented on src/Processors/QueryPlan/Optimizations/convertOuterJoinToInnerJoin.cpp:
```diff
@@ -99,14 +99,55 @@ size_t tryConvertAnyOuterJoinToInnerJoin(
+                const auto * current_node = nodes_to_process.top();
+                nodes_to_process.pop();
+
+                join_keys_interesting_side.insert(current_node->result_name);
```
I think we should add only `INPUT` nodes, shouldn't we? Example from test: `u1.name = reverse(u2.name)`.
It would contain: `reverse(u2.name)`, `u2.name`.
Would it be enough to have only `u2.name`?

antaljanosbenjamin replied:
I think you are right, it is enough to add `INPUT` nodes. My reasoning is if the node is not an input, then it is not in the grouping key for the group by. There might be a situation (`JoinStepLogical` -> `ExpressionStep` -> `Aggregating`) where the `ExpressionStep` might discard some columns, but I doubt we would discard a column that is used in the join. If this is the case, we can improve this later.

antaljanosbenjamin committed:
Do not convert joins in case of non equivalence conditions

antaljanosbenjamin committed:
Only check input nodes for checking the matching group

If there is a column used as grouping keys and in the join conditions, then it shouldn't be discarded in between and recalculated in the join.

antaljanosbenjamin committed:
Merge remote-tracking branch 'origin/master' into fix-convert-outer-to-inner-join-header-modification

antaljanosbenjamin committed:
Update expected output

antaljanosbenjamin commented:
I checked `TCP-H` `Q2`, `Q17` and `Q20`, the optimization still works.

novikd reviewed (approved):
LGTM
